src/integer/mat_poly_over_z/tensor.rs

```diff
@@ -0,0 +1,257 @@
+// Copyright © 2023 Marvin Beckmann
+//
+// This file is part of qFALL-math.
+//
+// qFALL-math is free software: you can redistribute it and/or modify it under
+// the terms of the Mozilla Public License Version 2.0 as published by the
+// Mozilla Foundation. See <https://mozilla.org/en-US/MPL/2.0/>.
+
+//! This module contains the implementation of the `tensor` product.
+
+use super::MatPolyOverZ;
+use crate::{
+    integer::PolyOverZ,
+    traits::{GetEntry, GetNumColumns, GetNumRows, Tensor},
+};
+use flint_sys::{fmpz_poly::fmpz_poly_mul, fmpz_poly_mat::fmpz_poly_mat_entry};
+
+impl Tensor for MatPolyOverZ {
+    /// Computes the tensor product of `self` with `other`
+    ///
+    /// Parameters:
+    /// - `other`: the value with which the tensor product is computed.
+    ///
+    /// Returns the tensor product of `self` with `other`.
+    ///
+    /// # Examples
+    /// ```
+    /// use qfall_math::integer::MatPolyOverZ;
+    /// use qfall_math::traits::Tensor;
+    /// use std::str::FromStr;
+    ///
+    /// let mat_a = MatPolyOverZ::from_str("[[1  1, 2  1 1]]").unwrap();
+    /// let mat_b = MatPolyOverZ::from_str("[[1  1, 1  2]]").unwrap();
+    ///
+    /// let mat_ab = mat_a.tensor_product(&mat_b);
+    /// let res_ab = "[[1  1, 1  2, 2  1 1, 2  2 2]]";
+    /// assert_eq!(mat_ab, MatPolyOverZ::from_str(res_ab).unwrap());
+    ///
+    /// let mat_ba = mat_b.tensor_product(&mat_a);
+    /// let res_ba = "[[1  1, 2  1 1, 1  2, 2  2 2]]";
+    /// assert_eq!(mat_ba, MatPolyOverZ::from_str(res_ba).unwrap());
+    /// ```
+    fn tensor_product(&self, other: &Self) -> Self {
+        let mut out = MatPolyOverZ::new(
+            self.get_num_rows() * other.get_num_rows(),
+            self.get_num_columns() * other.get_num_columns(),
+        );
+
+        for i in 0..self.get_num_rows() {
+            for j in 0..self.get_num_columns() {
+                let entry = self.get_entry(i, j).unwrap();
+
+                if !entry.is_zero() {
+                    unsafe { set_matrix_window_mul(&mut out, i, j, entry, other) }
+                }
+            }
+        }
+
+        out
+    }
+}
+
+/// This function sets a specific window of the provided matrix `out`
+/// according to the `scalar` multiple of `matrix`.
+///
+/// Sets the entries
+/// `[i*rows_other, j*columns_other]` up till `[row_left*(row_other +1), column_upper*(columns_other + 1)]`
+///
+/// Parameters:
+/// - `out`: the matrix in which the result is saved
+/// - `row_left`: defines the leftmost row of the set window
+/// - `column_upper`: defines the highest column of the set window
+/// - `scalar`: defines the value with which the part of the tensor product
+/// is calculated
+/// - `matrix`: the matrix with which the scalar is multiplied
+/// before setting the entries in `out`
+///
+/// Implicitly sets the entries of the matrix according to the definition
+/// of the tensor product.
+unsafe fn set_matrix_window_mul(
```

Comment:
Please add a `# Safety` section to the documentation describing why it is unsafe, i.e., matrix row_left and column_upper must be in a specific range.
If we have negative or too large `row_left` or `column_upper`, (really) bad stuff can happen. Randomly overwriting other data on the heap, and potentially also on the stack.

Maybe add an assert, however, since the function is private and unsafe that might not be necessary.

Reply:
`# Examples` is also missing, but since it is a private function just used in this file, I am fine without it.

Reply:
left out examples, added security section and asserts